lib/src/segy.c

```diff
@@ -1846,6 +1846,84 @@ int segy_traces( segy_datasource* ds,
     return SEGY_OK;
 }
 
+/* Allows only int2 and uint2. */
+static int field_data_to_scalar(
+    const segy_field_data* fd,
+    int* scalar
+) {
+    switch( fd->entry_type ) {
+        case SEGY_ENTRY_TYPE_INT2:
+            *scalar = fd->value.i16;
+            break;
+        case SEGY_ENTRY_TYPE_UINT2:
+            *scalar = fd->value.u16;
+            break;
+        default:
+            return SEGY_INVALID_FIELD_DATATYPE;
+    }
+    return SEGY_OK;
+}
+
+int segy_delay_recoding_time( segy_datasource* ds, float* delay ) {
+    char trace_header[SEGY_TRACE_HEADER_SIZE];
+
+    /* we don't need to figure out a trace size, since we're not advancing
+     * beyond the first header */
+    int err = segy_read_standard_traceheader( ds, 0, trace_header );
+    if( err != 0 ) {
+        return err;
+    }
+
+    const segy_entry_definition* standard_map =
+        ds->traceheader_mapping_standard.offset_to_entry_definition;
+    const uint8_t* standard_name_map =
+        ds->traceheader_mapping_standard.name_to_offset;
+
+    const int delay_offset = standard_name_map[SEGY_TR_DELAY_REC_TIME];
+    int delay_raw = 0;
+    segy_field_data delay_fd;
+    err = segy_get_tracefield(
+        trace_header, standard_map, delay_offset, &delay_fd
+    );
+    if( err != SEGY_OK ) return err;
+
+    // until someone complains, we restrict allowed types to these only
+    switch( delay_fd.entry_type ) {
+        case SEGY_ENTRY_TYPE_TIME2:
+        case SEGY_ENTRY_TYPE_INT2:
+            delay_raw = delay_fd.value.i16;
+            break;
+        case SEGY_ENTRY_TYPE_UINT2:
+            delay_raw = delay_fd.value.u16;
+            break;
+        default:
+            return SEGY_INVALID_FIELD_DATATYPE;
+    }
+
+    const int scalar_offset = standard_name_map[SEGY_TR_SCALAR_TRACE_HEADER];
+    int scalar_raw = 0;
+    segy_field_data scalar_fd;
+    err = segy_get_tracefield(
+        trace_header, standard_map, scalar_offset, &scalar_fd
+    );
+    if( err != SEGY_OK ) return err;
+
+    err = field_data_to_scalar( &scalar_fd, &scalar_raw );
+    if( err != SEGY_OK ) return err;
+
+    float delay_scalar;
+    if( scalar_raw == 0 ) {
+        delay_scalar = 1.0f;
+    } else if( scalar_raw < 0 ) {
+        delay_scalar = -1.0f / scalar_raw;
+    } else {
+        delay_scalar = 1.0f * scalar_raw;
+    }
```

Comment:
We probably don't need it here yet, but maybe this could become its own method. I assume that is the reconstruction of a `time2` value. In this sense, it could be reused other places later.